File: java/bob/src/main/java/Bob.java
import java.util.regex.Pattern;

public class Bob {
    public String hey(String something) {
        Phrase phrase = new Phrase(something);
        if (phrase.isSilence()) return "Fine. Be that way!";
        if (phrase.isShout()) return "Whoa, chill out!";
        if (phrase.isQuestion()) return "Sure.";
        return "Whatever.";
    }

    public class Phrase {
        private final Pattern UPPERCASE_MATCHER = Pattern.compile(".*\\p{Upper}.*", Pattern.UNICODE_CHARACTER_CLASS);
        private final Pattern LOWERCASE_MATCHER = Pattern.compile(".*\\p{Lower}.*", Pattern.UNICODE_CHARACTER_CLASS);
        private final String phrase;

        public Phrase(String phrase) {
            this.phrase = phrase;
        }

        public boolean isSilence() {
            return phrase.replaceAll("\\s", "").isEmpty();
        }

        public boolean isShout() {
            return hasUppercaseLetters() && !hasLowercaseLetters();
        }

        public boolean isQuestion() {
            return phrase.endsWith("?");
        }

        private boolean hasLowercaseLetters() {
            return LOWERCASE_MATCHER.matcher(phrase).find();
        }

        private boolean hasUppercaseLetters() {
            return UPPERCASE_MATCHER.matcher(phrase).find();
        }
    }
}

File: go/bob/bob.go
package bob

import (
	"regexp"
	"strings"
)

const testVersion = 2

func Hey(phrase string) string {
	greeting := Greeting{phrase: phrase}

	if greeting.IsShout() {
		return "Whoa, chill out!"
	}
	if greeting.IsQuestion() {
		return "Sure."
	}
	if greeting.IsSilence() {
		return "Fine. Be that way!"
	}
	return "Whatever."
}

type Greeting struct {
	phrase string
}

func (g Greeting) IsShout() bool {
	return g.containsUppercase() && g.allUppercase()
}

func (g Greeting) IsQuestion() bool {
	endsWithQuestionMark := `\?\s*\z`
	return g.matches(endsWithQuestionMark)
}

func (g Greeting) IsSilence() bool {
	allNonWordCharacters := `\A\W*\z`
	return g.matches(allNonWordCharacters)
}

func (g Greeting) containsUppercase() bool {
	return g.matches(`[A-Z]`)
}

func (g Greeting) allUppercase() bool {
	return g.phrase == strings.ToUpper(g.phrase)
}

func (g Greeting) matches(pattern string) bool {
	matcher := regexp.MustCompile(pattern)
	return matcher.Match([]byte(g.phrase))
}
